### ISSUE TITLE
Multiple Cameras, One Viewport

### ISSUE THREAD

**Issue by Mantissa-23** [godotengine/godot-proposals#956]
**Describe the project you are working on:**
A first-person shooter, making use of first-person viewmodels that are separate from third-person/worldspace models.

**Describe the problem or limitation you are having in your project:**
Godot currently only supports rendering one camera per viewport. This makes the following use-cases difficult to implement:
- Secondary camera with fixed FOV for consistent viewmodel appearance despite changes to primary camera FOV
- 3d "skybox" implementation where distant/background objects are rendered at a smaller scale in a separate scene.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
Allowing multiple cameras (with different render masks) to render to the same viewport would overcome this problem. For each use case:
- A primary camera is used to render the world, with player-changeable FOV. A secondary camera is used to render viewspace objects, such as gun viewmodels, hands, grenades, etc.
- A primary camera is used to render near objects; a secondary camera is used to render far objects which are stored in a separate 3d render layer.

This is a feature that is present in both Unreal and Unity, and is used for similar reasons.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
Cameras will have a `depth` field, which determines the order in which they render; pixels that reach the background of one camera (i.e. are not rendered) at a higher depth will be rendered by the next lowest camera.

I don't have an understanding of Godot's rendering pipeline, and understand it is in flux for Godot 4.0. I have C++ experience, and with some pointers I'd be willing to implement this myself for 3.2.2 or 4.0.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
I believe these can currently be implemented using a screen-space ViewportTexture and Viewport nodes, however, I haven't tested this, and I imagine the performance would be worse than rendering to the same, default Viewport from multiple cameras. If the performance would be similar, go ahead and close this issue.

That being said, I do believe using ViewportTextures is still less _intuitive_ than having multiple cameras rendering to a single Viewport.

**Is there a reason why this should be core and not an add-on in the asset library?:**
This requires a change to the Camera node as well as how it interfaces with the rendering pipeline. I also cannot imagine it being a large change, either code-wise or size-wise.

**Comment by 31** [comment]
This is similar to what I've been thinking about at https://github.com/godotengine/godot-proposals/issues/932! Your scenario (FPS viewmodels) is **way** less obscure than mine though, and shows IMO that some sort of control over object render order and depth buffer manipulation is needed in the 4.0 pipeline configurability plans.

**Comment by Flarkk** [comment]
I guess this is also related to #355 in some way (panoramic skybox)

**Comment by DylanCulfogienis** [comment]
@Flarkk yup, I'm biased but I'd argue that this approach is more generic, allowing for both foreground and background scenes.

**Comment by Chatto** [comment]
I second(or third) this proposal. It would allow me to make some ground up hyperspace visuals(Ever smoke DMT?) I could also then use this procedurally as well.

**Comment by 31** [comment]
I tried my hand at this and got a proof of concept working on 3.2, with bugs. I opened https://github.com/godotengine/godot-proposals/issues/1428 with more details. If someone wants to try it out, work on it, take it over, etc., would be appreciated. Opened it separately because it's a particular way of implementing this and don't want to sidetrack this thread. (But I do want to let anyone subscribing know, if you're interested. 😄)

**Comment by aaronfranke** [comment]
See also: https://github.com/godotengine/godot/pull/8215

**Comment by Yazir** [comment]
Stumbled upon this proposal while searching for an answer to the fps viewmodel separate camera question. Here's a way I've managed to make it work in my project.

![obraz](https://user-images.githubusercontent.com/8749862/91841984-81106f00-ec53-11ea-9082-38f0e7bce88b.png)
2 viewports overlayed on top of each other, the second viewport has **Transparent Bg** set to true.
Cameras are using separate cull layers (viewmodel uses the same cull layer as the viewmodel camera)

A video of the effect in action
https://youtu.be/RLM0oX2eptg

I believe it is fairly easy to setup and could be hopefully of use to you :)

**Comment by 31** [comment]
> 2 viewports

In this proposal, the concern with that approach is that it may be inefficient. I use this too as a workaround for now. 🙂 

The roadblock for me is that this doesn't work in VR. You probably wouldn't want a viewmodel in VR, but here are other effects that would be solved by this proposal.

**Comment by Yazir** [comment]
If you render only what's in the mask of the viewmodel camera, why would it be not performant? You don't double render objects. There should be some performance test done.

**Comment by Calinou** [comment]
> Secondary camera with fixed FOV for consistent viewmodel appearance despite changes to primary camera FOV

This can be simulated fairly well by using non-uniform scaling on the viewmodel. Stretch it to make the FOV look higher, and squash it to make the FOV look lower. If you run into mesh normal issues in `3.x`, enable **Ensure Correct Normals** in the viewmodel's SpatialMaterial.

**Comment by farfalk** [comment]
Is there any update on this? I'm trying to implement this approach https://youtu.be/LAopDQDCwak in Godot - but the impression I get from this issue and other related ones is that it's simply not possible

**Comment by pycbouh** [comment]
> Is there any update on this? I'm trying to implement this approach https://youtu.be/LAopDQDCwak in Godot - but the impression I get from this issue and other related ones is that it's simply not possible

You can have multiple viewports. This proposal mentions that it may not be the most efficient way, but it is still a way.

**Comment by farfalk** [comment]
> > Is there any update on this? I'm trying to implement this approach https://youtu.be/LAopDQDCwak in Godot - but the impression I get from this issue and other related ones is that it's simply not possible
> 
> You can have multiple viewports. This proposal mentions that it may not be the most efficient way, but it is still a way.

Yeah I managed to get it working via two different viewports, each one with a camera, one of them with transparent_bg enabled, both represented on two overlapped TextureRects as ViewportTextures.

The problem I'm encountering right now is that the transparent_bg'd one isn't affected by the environment of its camera. I'll try to  look for / open an issue about it in the next days

**Comment by djmaesen** [comment]
thumbs up for this feature

**Comment by Calinou** [comment]
@djmaesen Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by elvisish** [comment]
Has this had any traction since May? It's certainly unusual there's no alternative to using either an FOV/forced depth shader or shrinking the FPS weapon models down to 0.01. I think multiple cameras on the same viewport is probably the most Godot-friendly way of doing this.

**Comment by Calinou** [comment]
> Has this had any traction since May? It's certainly unusual there's no alternative to using either an FOV/forced depth shader or shrinking the FPS weapon models down to 0.01. I think multiple cameras on the same viewport is probably the most Godot-friendly way of doing this.

As far as I know, nobody made any progress on implementing this (or even started looking into it). It's not a planned feature for 4.0.

**Comment by Mantissa-23** [comment]
Yeah for now I'm using @Yazir's solution, which is currently the only way of achieving this effect in Godot.

I have no idea if it would be more or less efficient to do this using multiple cameras per viewport. Currently I have 3 of them, one for rendering a viewmodel, one for rendering a 3D HUD, and then one for rendering the actual game itself.

If it isn't necessarily more efficient to render multiple cameras to a single viewport (i.e. multiple viewports have performance on the same order as one with multiple cameras), then I'd argue that this proposal isn't relevant any longer and could be closed, as long as @31's issue with this particular solution not working in VR is also resolved in some way.

If someone with more knowledge of the rendering pipeline could key in here, we might be able to close this.

**Comment by elvisish** [comment]
> Yeah for now I'm using @Yazir's solution, which is currently the only way of achieving this effect in Godot.
> 
> 
> 
> I have no idea if it would be more or less efficient to do this using multiple cameras per viewport. Currently I have 3 of them, one for rendering a viewmodel, one for rendering a 3D HUD, and then one for rendering the actual game itself.
> 
> 
> 
> If it isn't necessarily more efficient to render multiple cameras to a single viewport (i.e. multiple viewports have performance on the same order as one with multiple cameras), then I'd argue that this proposal isn't relevant any longer and could be closed, as long as @31's issue with this particular solution not working in VR is also resolved in some way.
> 
> 
> 
> If someone with more knowledge of the rendering pipeline could key in here, we might be able to close this.
> 
> 

It's awkward though using multiple viewports compared to just adding another camera, and I seem to recall lighting didn't match using a second viewport.

**Comment by Yazir** [comment]
@elvisish 
It is awkward, true, but as I recall, lighting worked properly, as seen in the video in my comment above from 2020. Afair it can be turned off and on, depending on light masks.

**Comment by elvisish** [comment]
This really should be possible in Godot 4.0, at the very least planned for 4.1 as it's incredibly cumbersome to do without being able to just show certain objects on certain cameras. Maybe even a 3D Objects Layer node would be better? Is there a way of using 4.0's SubViewport to do this easily?

**Comment by aaronfranke** [comment]
@elvisish 4.0 is already out and 4.1 will be out soon (feature freeze in 2 weeks), so there isn't really time to do this in 4.1.

**Comment by elvisish** [comment]
> @elvisish 4.0 is already out and 4.1 will be out soon (feature freeze in 2 weeks), so there isn't really time to do this in 4.1.

Ahh I didn't realize it was this close, maybe 4.2?

**Comment by Calinou** [comment]
While this is a heavily desired feature, we still don't have a clear idea of how to implement this, and nobody has really looked into it so far. There are many difficult problems to resolve, such as handling GI interactions when using multiple cameras. You generally expect the weapon model to be able to receive GI from the scene after all, and this is made more difficult when the second camera uses a different FOV.

**Comment by 0xafbf** [comment]
Coincidentially I faced a related issue today, so I think it is worth to add my two cents: 

A colleague is currently working on some atmospheric effects, following this video: https://www.youtube.com/watch?v=dzcFB_9xHtg , all of this uses custom shaders for the atmosphere.

After following the video (and using the project author's code), we are using multiple camera+viewport. The first one is for drawing the stars in the background, which is convenient to use some reduced units (like, not hundreds of thousands of kilometers away from the camera. Then we use another one for the planet, which draws on top with alpha blending. And in the third one we draw the atmosphere. 

The linked project actually draws the stars last, at first we didn't understand why, but after trying to move them to the back of the stack, we found out why. The author blends the atmosphere with the globe by grabbing the globe camera render target and using it as a parameter for the atmosphere shader that renders in the atmosphere camera+viewport. when he does this, he is not able to blend the stars together because they were in a different viewport to begin with, so to adjust for that, he would have needed to change the atmosphere shader and make it more complex

Something we thought on doing, but couldn't figure a way to do, is to set the viewport blend mode as additive, this could have helped us but I think it still wouldn't have been the right solution. If we did it that way, post-processing effects would apply separately and then the rest of the image processing would be incorrect. 

It makes sense to compile the effect of many cameras into a single viewport because then all draw operations would happen in the same buffer and tonemapping would occur once. For now we are moving everything to a single camera and reprojecting the stars meshes in the shader. We're hoping that doing it this way would blend out the stars naturally when you get inside the atmosphere as the atmosphere brightness would affect the exposure, and make the stars invisible. We could always add many post-process steps later, as long as the pre-tonemapped result is accurate.

**Comment by 31** [comment]
> as long as @31's issue with this particular solution not working in VR is also resolved in some way.

I believe the VR integration could be independently improved such that multiple viewports with per-eye rendering would work, but (with no real knowledge of the plans of everyone involved!) I imagine it would be much less likely for that to be implemented given the relatively small number of VR headsets vs. tagging along with a more broadly requested new Godot feature that applies to many FPS games. 😄 

I did briefly look at improving the VR handling, but there are many intertwined parts so I gave up and worked on the implementation I posted in https://github.com/godotengine/godot-proposals/issues/1428 instead.

**Comment by 2fd5** [comment]
> I believe it is fairly easy to setup and could be hopefully of use to you :)

Not sure how does the camera movement work for you with that solution. I am using Godot 4.1beta3 and with same setup where Camera2d is a child of SubViewport my camera movement sudently have no effect, SubViewportContainer just displays what is visible from 0,0 to the size of SubVieport node. 
Was it a project in Godot 3 ?

For my particular use case I just want to render part of the scene to a separate texture and then display it with some shader doing masking applied, with important aspect being that I only render what Camera sees. I came here as hoped for having multiple cameras transformed in same way but rendering separate Visibility Layers. Or having Camera2D have multiple render layers or render passes that I could use with different shaders.  

I have not found a way to achieve that. Hope that will some more use cases when the work on this FR starts. Thanks

**Comment by Flarkk** [comment]
#7916 may help out here

**Comment by digi-chris** [comment]
I wanted to add something here because I'm using viewports very heavily to do a lot of what's been suggested above, especially with regards to drawing very far away objects (stars), then planets, atmosphere, far away terrain and finally near terrain. The only way I can get this working well currently is with multiple viewports stacked on top of each other.

I'm facing three main issues that I don't think have been mentioned in this thread (sorry if I've missed something!):

- Setting up these viewports seems to add a fairly substantial RAM overhead... when I remove the viewports and have everything rendered with a single camera, the RAM usage at startup drops by 3gb. This seems a little wasteful when I know with the right tools I could just part render, clear the depth buffer and switch cameras, then continue rendering to the same viewport. Unfortunately I can't just drop to one camera because I run into overflow issues, and my LOD terrain process works so much better when I can separate things out with different cameras.
- Effects like FSR don't work with overlaid viewports, because they lose their transparent backgrounds when you switch that effect on - so for example if the topmost viewport has FSR enabled, then all the other viewports disappear. This is limiting my ability to get good visuals in game.
- Screen-space shaders become a problem when you have multiple viewports, because you can only modify the viewport you're working on, with no knowledge of the previous viewports. Say you want to do some sort of sunlight/flare effect, you're going to have to run the same effect on every single viewport rather than just stacking it on top at the end, and it won't necessarily blend together very well.

I've tried a number of different workarounds that might improve performance or even get the visuals looking closer to what I can acheive with a single camera, but I keep coming back to thinking I should be able to use more than one camera within a viewport - it would help me out a lot!

What I'd love to be able to do is start the render with one camera, then clear the depth buffer and switch to another camera, and repeat for as many cameras as I have. Maybe you could use the layer masks to define what gets rendered with each camera, and just order it based on the order the cameras appear in the tree?

Even better would be if I could define max/min depth values for each camera and then not have to clear the depth buffer between switching cameras.

**Comment by Calinou** [comment]
> especially with regards to drawing very far away objects (stars), then planets, atmosphere, far away terrain and finally near terrain. 

Note that https://github.com/godotengine/godot/pull/99986 enables this use case without requiring the use of multiple viewports, even in single-precision builds.

**Comment by digi-chris** [comment]
> Note that [godotengine/godot#99986](https://github.com/godotengine/godot/pull/99986) enables this use case without requiring the use of multiple viewports, even in single-precision builds. This is possible thanks to the reverse Z buffer, although you need to use the Forward+ rendering method to make full use of it (since it uses a more precise depthbuffer than Mobile and Compatibility).

This looks amazing, and would certainly solve a number of issues I'm currently facing. Reading through the PR, the only issue that would cause me some problems is that it looks like SDFGI crashes with large znear or zfar values. If that can be solved, it'd work for me.

**Comment by Flarkk** [comment]
For later reference I just fixed the crash in #99986 with SDFGI and large zfar values (see #104120).

**Comment by 31** [comment]
I don't see it being linked yet, so thought I'd point out another proposal for a more directly customizable render pipeline. Something for those of us who want this feature for more than just z-buffer accuracy mitigation:

* https://github.com/godotengine/godot-proposals/issues/644

(I think the best summary is this 2020 reduz comment: https://github.com/godotengine/godot-proposals/issues/644#issuecomment-616532981.)

**Comment by core6quad** [comment]
any updates on this?

rendering viewmodels in a seperate SubViewport has a bunch of limitations like light not being blocked by level's geometry (light being unaffected by geomtery from an another layer)

**Comment by Calinou** [comment]
> rendering viewmodels in a seperate SubViewport has a bunch of limitations like light not being blocked by level's geometry (light being unaffected by geomtery from an another layer)

For viewmodels, you should use [Z Clip Scale](https://docs.godotengine.org/en/latest/tutorials/3d/standard_material_3d.html#use-z-clip-scale) and optionally [Use FOV Override](https://docs.godotengine.org/en/latest/tutorials/3d/standard_material_3d.html#use-fov-override). This will be much faster than using a separate viewport, while keeping rendering features intact.

**Comment by paulovbettio** [comment]
> For viewmodels, you should use [Z Clip Scale](https://docs.godotengine.org/en/latest/tutorials/3d/standard_material_3d.html#use-z-clip-scale) and optionally [Use FOV Override](https://docs.godotengine.org/en/latest/tutorials/3d/standard_material_3d.html#use-fov-override). This will be much faster than using a separate viewport, while keeping rendering features intact.

In my experience, viewmodels with multiple materials and parts gets overwhelming very fast on fov override and z clip scale setups.

A good solution to this would be the possibility of setting a material overlay for the parent Node3D and all their children inheriting it, but I'm not sure how to handle this with VisualInstance3D... Maybe a VisualGroup3D node?
